eva_submission/biosamples_submission.py

```diff
@@ -0,0 +1,471 @@
+            if 'last' in json_response['_links']: json_response['_links'].pop('last')
+            if 'next' in json_response['_links']: json_response['_links'].pop('next')
+        return json_response
+
+    def follows_link(self, key, json_obj=None, method='GET', url_template_values=None, join_url=None, **kwargs):
+        """
+        Same function as follows but construct the query_string from a single keyword surrounded by '_links' and 'href'.
+        """
+        return self.follows(('_links', key, 'href'),
+                            json_obj=json_obj, method=method, url_template_values=url_template_values,
+                            join_url=join_url, **kwargs)
+
+    @cached_property
+    def root(self):
+        return self._req('GET', self.bsd_url).json()
+
+
+class BSDSubmitter(AppLogger):
+
+    def __init__(self, communicator, domain):
+        self.communicator = communicator
+        self.domain = domain
+        self.sample_name_to_accession = {}
+
+    def validate_in_bsd(self, samples_data):
+        for sample in samples_data:
+            sample['domain'] = self.domain
+            self.communicator.follows_link('samples', join_url='validate', method='POST', json=sample)
+
+    def submit_to_bsd(self,  samples_data):
+        """
+        This function creates or updates samples in BioSamples and return a map of sample name to sample accession
+        """
+
+        for sample in samples_data:
+            sample['domain'] = self.domain
+            if 'accession' not in sample:
+                # Create a sample
+                sample_json = self.communicator.follows_link('samples', method='POST', json=sample)
+                self.debug('Accession sample ' + sample.get('name') + ' as ' + sample_json.get('accession'))
+            else:
+                # Update a sample
+                self.debug('Update sample ' + sample.get('name') + ' with accession ' + sample.get('accession'))
+                sample_json = self.communicator.follows_link('samples', method='PUT', join_url=sample.get('accession'),
+                                                             json=sample)
+            self.sample_name_to_accession[sample_json.get('name')] = sample_json.get('accession')
+
+
+class SampleSubmitter(AppLogger):
+
+    sample_mapping = {}
+
+    project_mapping = {}
+
+    def __init__(self):
+        communicator = HALCommunicator(cfg.query('biosamples', 'aap_url'), cfg.query('biosamples', 'bsd_url'),
+                                       cfg.query('biosamples', 'username'), cfg.query('biosamples', 'password'))
+        self.submitter = BSDSubmitter(communicator, cfg.query('biosamples', 'domain'))
+
+    @staticmethod
+    def map_key(key, mapping):
+        """
+        Retrieve the mapping associated with the provided key or returns the key if it can't
+        """
+        if key in mapping:
+            return mapping[key]
+        else:
+            return key
```

Comment:
This method return seems to be missing the ".lower()" method as found [here](https://github.com/EBIvariation/metadata/blob/3bc04293bc2508060fc9b54ad27b9e92d3158bba/submission_to_biosamples/submit_to_bsd.py#L207).

Reply:
Yes I moved it to make it more explicit.